index.js

```diff
@@ -137,11 +138,58 @@ class Installer {
           this.pkgCount++
           return this
         })
+        .catch(e => {
+          if (child.optional) {
+            this.pkgCount++
+            child.failed = true
```

Comment:
Rather than marking this as failed via a property, could you do a Set for failed here? I know that in the mark phase, you’re iterating over the top level dependencies only to find failed deps, so a Set doesn’t save *much*, but it’s something.